src/neuralnet/superblock.h

```diff
@@ -0,0 +1,741 @@
+
+        IMPLEMENT_SERIALIZE
+        (
+            READWRITE(VARINT(m_total_credit));
+            READWRITE(VARINT(m_average_rac));
+            READWRITE(VARINT(m_rac));
+        )
+    };
+
+    //!
+    //! \brief An optional type that either contains some project statistics or
+    //! does not.
+    //!
+    typedef boost::optional<ProjectStats> ProjectStatsOption;
+
+    //!
+    //! \brief Contains aggregated project statistics.
+    //!
+    class ProjectIndex
+    {
+    public:
+        typedef std::map<std::string, ProjectStats>::size_type size_type;
+        typedef std::map<std::string, ProjectStats>::iterator iterator;
+        typedef std::map<std::string, ProjectStats>::const_iterator const_iterator;
+
+        //!
+        //! \brief Initialize an empty project index.
+        //!
+        ProjectIndex();
+
+        //!
+        //! \brief Returns an iterator to the beginning.
+        //!
+        const_iterator begin() const;
+
+        //!
+        //! \brief Returns an iterator to the end.
+        //!
+        const_iterator end() const;
+
+        //!
+        //! \brief Get the number of projects contained in the superblock.
+        //!
+        size_type size() const;
+
+        //!
+        //! \brief Determine whether the index contains any projects.
+        //!
+        bool empty() const;
+
+        //!
+        //! \brief Get the sum of the recent average credit of all the projects
+        //! on the whitelist.
+        //!
+        //! \return Total project RAC at the time of the superblock.
+        //!
+        uint64_t TotalRac() const;
+
+        //!
+        //! \brief Get the average recent average credit of all the projects on
+        //! the whitelist.
+        //!
+        //! \return Average project RAC at the time of the superblock.
+        //!
+        double AverageRac() const;
+
+        //!
+        //! \brief Try to get the project statistics for a project with the
+        //! specified name.
+        //!
+        //! \param name As it exists in the current whitelist.
+        //!
+        //! \return An object that contains the matching project statistics if
+        //! it exists.
+        //!
+        ProjectStatsOption Try(const std::string& name) const;
+
+        //!
+        //! \brief Add the supplied project statistics to the index.
+        //!
+        //! This method ignores an attempt to add a duplicate entry if a project
+        //! already exists with the same name.
+        //!
+        //! \param name  As it exists in the current whitelist.
+        //! \param stats Contains project RAC data.
+        //!
+        void Add(std::string name, const ProjectStats& stats);
+
+        IMPLEMENT_SERIALIZE
+        (
+            READWRITE(m_projects);
+
+            // Tally up the recent average credit after deserializing.
+            //
+            if (fRead) {
+                REF(m_total_rac) = 0;
```

Comment:
What is the `REF` for here?

Reply:
Yeah, the statements in `IMPLEMENT_SERIALIZE` are also expanded into the `Serialize()` and `GetSerializeSize()` methods which have `const` qualifiers. Even though we only run the code above in `Unserialize()`, we need the `REF` to do a `const_cast` so that the compiler allows it.